internals/overlord/logstate/loki/loki.go

```diff
@@ -79,18 +86,25 @@ func fillDefaultOptions(options *ClientOptions) *ClientOptions {
 
 func (c *Client) Add(entry servicelog.Entry) error {
 	if n := len(c.entries); n >= c.options.MaxRequestEntries {
-		// Buffer full - remove the first element to make room
+		// 'entries' is full - remove the first element to make room
 		// Zero the removed element to allow garbage collection
 		c.entries[0] = lokiEntryWithService{}
 		c.entries = c.entries[1:]
 	}
-	if cap(c.entries)-len(c.entries) == 0 {
-		// Allocate a new slice with capacity equal to double the buffer size, and
-		// copy over the currently buffered logs
-		c.entries = append(
-			make([]lokiEntryWithService, 0, 2*c.options.MaxRequestEntries),
-			c.entries...)
+
+	if len(c.entries) >= cap(c.entries) {
+		// Copy all the elements to the start of the buffer
+		copy(c.buffer, c.entries)
+
+		// Reset the view into the buffer
+		c.entries = c.buffer[0:len(c.entries):len(c.buffer)]
+
+		// Zero removed elements to allow garbage collection
+		for i := len(c.entries); i < len(c.buffer); i++ {
```

Comment:
`clear(c.buffer[len(c.entries):])`?

Reply:
Unfortunately `clear()` is only available in Go 1.21, which we're not on yet.

Reply:
that's nice, unfortunately `clear` was added in Go 1.21, while Pebble is still on Go 1.20